### PR TITLE
Open examples in CodeSandbox

### LINKED ISSUE #2126: Open examples in CodeSandbox
# 🙋 Documentation Request

It would be very convenient to be able to open any example from the docs in its own CodeSandbox. We should be able to achieve this using CodeSandbox’s [API](https://codesandbox.io/docs/api), and do so without the need to maintain these examples separately from the ones being rendered in the docs. There could be a link near each example that opens a newly created CodeSandbox in a new tab, and includes relevant packages with correct versions and imports.

Screen Shot 2021-07-20 at 4 47 45 PM

## Appearance

Examples should feature a link that is labeled "Open Sandbox" and is positioned in the top right corner of code examples.

## Behavior

When the link is clicked, a new tab should open with a newly created CodeSandbox that includes the relevant example code. The CodeSandbox should include

- `title`: ComponentName v3.x.x (optional subheader) - React Spectrum
- `description`: ComponentName example from Adobe's React Spectrum library
- `tags`: react, react-spectrum, react-aria, adobe

and `files` with
- `package.json`: includes latest versions for `react`, `react-dom`, `@adobe/react-spectrum`, along with all all components used in example code
- `Example.js`: example code
- `index.js`: renders `<Example />` wrapped in `Provider`
- `index.html`: standard from CodeSandbox React template

We should also properly handle any examples that don't include any rendering (i.e. many from `react-stately` docs).

## Other Thoughts

- Since the examples in the docs may eventually feature tabs or a toggle for showing TypeScript code examples, we should keep that in case it affects this UI decision. (#865) 
- Another UI option could be to have a CodeSandbox “tab” on examples that shows a newly created sandbox embedded within the docs page via an iframe.

## 🧢 Your Company/Team

RSP

### PR DESCRIPTION
**Pull request by reidbarber**
Closes #2126 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Open docs, find "Open Sandbox" buttons next to example code.

## 🧢 Your Project:

RSP

### CODE REVIEW

**Comment by reidbarber** [comment]
Starting this in order to get some early feedback. It seems like we could do a lot of the work for this either on the client or at build time. I think doing as much as possible on the client is probably preferable as to not inflate the build, but at the same time work at build time might be more maintainable and also give us access to more info (some examples that import other components). Currently everything is done on the client, but we should probably move a bit of the work to the build. I’m just getting familiar with the docs build, so feel free to recommend better methods for implementing this where you see fit. 

Still need to:
- [ ] Handle examples that include other components that aren’t explicitly imported in the example
- [ ] Handle examples that don’t do any rendering (i.e. many from react-stately)
- [ ] Add more specific titles for sandboxes (i.e. Button (Disabled))

File: packages/dev/docs/src/createCodeSandbox.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  <Provider theme={defaultTheme}>
+    <Example />
+  </Provider>,
+  rootElement
+);
+`;
+
+export function createCodeSandbox(e) {
+  // Disable button while awaiting sandbox creation.
+  e.currentTarget.disabled = true;
```
Comment (LFDanLu) [inline packages/dev/docs/src/createCodeSandbox.js:74]: I noticed that the button remains disabled after the sandbox is created, is this intentional?

File: packages/dev/docs/src/Layout.js [hunk 2]
```diff
@@ -68,7 +68,14 @@ const mdxComponents = {
   p: ({children, ...props}) => <p className={typographyStyles['spectrum-Body3']} {...props}>{children}</p>,
   ul: ({children, ...props}) => <ul {...props} className={typographyStyles['spectrum-Body3']}>{children}</ul>,
   ol: ({children, ...props}) => <ol {...props} className={typographyStyles['spectrum-Body3']}>{children}</ol>,
-  code: ({children, ...props}) => <code {...props} className={typographyStyles['spectrum-Code4']}>{children}</code>,
+  code: ({children, ...props}) => (
+    <>
+      <button className={clsx(buttonStyles['spectrum-ActionButton', 'spectrum-ActionButton--sizeM', 'spectrum-ActionButton--quiet'], 'codeSandboxButton')} style={{float: 'right'}}>
+        <span className={buttonStyles['spectrum-ActionButton-label']}>Open Sandbox</span>
+      </button>
```
Comment (LFDanLu) [inline packages/dev/docs/src/Layout.js:75]: IMO this button could use a more descriptive `aria-label` (maybe "Open current example in a code sandbox"?), currently just states "Open Sandbox button" which might be confusing w/o context

File: packages/dev/docs/src/createCodeSandbox.js (hunk 1 above)
Comment (reidbarber) [reply packages/dev/docs/src/createCodeSandbox.js]: Changed to a link now :)

File: packages/dev/docs/src/Layout.js (hunk 2 above)
Comment (reidbarber) [reply packages/dev/docs/src/Layout.js]: Fixed, thanks!

File: packages/dev/docs/src/createCodeSandbox.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  // Remove imports from example and add to page's imports
+  exampleCode = exampleCode.replace(importsRegex, (m, _, s) => {
+    pageImports += m;
+    return '';
+  });
+
+  pageImports.replace(importsRegex, (m, _, s) => {
+    let importGroup = _.replace(/[{}\s]/g, '').split(',');
+    importGroup.forEach(i => {
+      // For now, we'll include any import if the name is present. This could give some false positives, but it should ensure no missing imports.
```
Comment (snowystinger) [inline packages/dev/docs/src/createCodeSandbox.js:107]: Yep, lets add it to things to do in followup

**Review by snowystinger — COMMENTED** [review]
This is so cool!

File: packages/dev/docs/src/client.js [hunk 4]
```diff
@@ -217,3 +219,20 @@ window.addEventListener('pagehide', () => {
...
+
+// Only add CodeSandbox links to react-spectrum examples now. We'll add react-aria and react-stately support later
+if (window.location.pathname.includes('/react-spectrum/')) {
+  [...document.querySelectorAll('code[metastring^="example"]')].forEach(element => {
+    let div = document.createElement('span');
+    div.classList.add('codeSandboxLink');
+    element.prepend(div);
+  
+    ReactDOM.render(
+      <Link aria-label="Open current example in a code sandbox" UNSAFE_style={{float: 'right'}} onPress={createCodeSandbox}>
```
[the remark this answers is not in the document]
Comment (reidbarber) [reply packages/dev/docs/src/client.js]: Agreed, and I think it looks better as well. Went ahead and changed it for now so there is a release to compare.

**Comment by reidbarber** [comment]
All included links should now create working sandboxes. Currently, these just show on React Spectrum examples (including pre-release components), but we can add to React Aria and React Stately examples later.

I introduced two options that can be added to examples to override the default functionality regarding sandboxes:
- add `sandbox-disabled=true` to an example to not include a sandbox link
- add `fragment=true` to examples that are fragments in order to wrap them in the sandbox (Used by some examples in LogicButton, Form, ProgressCircle, StatusLight).

Would love to hear any feedback on these options, and if this is something that we would want to do or not, or if there are better solutions.

There are some examples that have sandboxes disabled in this PR, notably ones on pages in the Introduction/Concepts sections and a few that weren't quite working yet (on TableView, DialogContainer, and ActionButton pages).

File: packages/dev/docs/src/createCodeSandbox.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  if (/^\s*function (.|\n)*}\s*$/.test(exampleCode)) {
+    exampleCode = `import React from 'react';
+${imports}
+export default ${exampleCode}
+`;
+  } else {
+    let nonRenderCode = exampleCode.replace(/^(<(.|\n)*>)$/m, () => '');
+    exampleCode = exampleCode.replace(nonRenderCode, '').trim();
+    let extraIncluded = false;
+    extraCode.replace(/(let\s|const\s|var\s|function\s)\w+/g, (a) => {
```
Comment (snowystinger) [inline packages/dev/docs/src/createCodeSandbox.js:146]: What is extra code? example in a comment could help as well

File: packages/dev/parcel-transformer-mdx-docs/MDXTransformer.js [hunk 6]
```diff
@@ -45,6 +47,11 @@ module.exports = new Transformer({
           let [meta, ...options] = (node.meta || '').split(' ');
           if (meta === 'import') {
             exampleCode.push(node.value);
+            exampleImports.push(node.value);
+            let nonImportCode = node.value.replace(/import ((?:.|\n)*?) from (['"].*?['"]);?/g, '').trim();
```
Comment (snowystinger) [inline packages/dev/parcel-transformer-mdx-docs/MDXTransformer.js:51]: lets get comments on all of the regex's so I don't have to try to decipher them

**Comment by snowystinger** [comment]
I think my overall comment is, if i forget any of the extra `sandbox-disabled` or `fragment`, am I liable to create links to bad examples in our docs? If that's the case, we should probably aim for an opt in approach instead of an opt out, what do others think?

**Comment by reidbarber** [comment]
Closing, may reopen at some point.

**Pull request closed without merging**